pkg/syncer/controller/controller.go

```diff
@@ -0,0 +1,186 @@
+func NewSyncConfigController(
+	logger klog.Logger,
+	syncConfigClient edgev1alpha1typed.EdgeSyncConfigInterface,
+	syncConfigInformer edgev1alpha1informers.EdgeSyncConfigInformer,
+	syncConfigName string,
+	upSyncer syncers.SyncerInterface,
+	downSyncer syncers.SyncerInterface,
+) (*controller, error) {
+	rateLimitter := workqueue.NewItemFastSlowRateLimiter(5*time.Second, 30*time.Second, 1000)
+	c := &controller{
+		queue:                       workqueue.NewNamedRateLimitingQueue(rateLimitter, controllerName),
+		syncConfigLister:            syncConfigInformer.Lister(),
+		syncConfigInformerHasSynced: syncConfigInformer.Informer().HasSynced,
+		syncConfigClient:            syncConfigClient,
+		upSyncer:                    upSyncer,
+		downSyncer:                  downSyncer,
+	}
+	syncConfigInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err != nil {
+				return false
+			}
+			_, name, err := cache.SplitMetaNamespaceKey(key)
+			if err != nil {
+				return false
+			}
+			return name == syncConfigName
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { c.enqueueSyncConfig(obj, logger) },
+			UpdateFunc: func(old, obj interface{}) { c.enqueueSyncConfig(obj, logger) },
+			DeleteFunc: func(obj interface{}) { c.enqueueSyncConfig(obj, logger) },
+		},
+	})
+
+	return c, nil
+}
+
+// controller is a control loop that watches EdgeSyncConfig.
+type controller struct {
+	queue                       workqueue.RateLimitingInterface
+	syncConfigLister            edgev1alpha1listers.EdgeSyncConfigLister
+	syncConfigInformerHasSynced cache.InformerSynced
+	syncConfigClient            edgev1alpha1typed.EdgeSyncConfigInterface
+	upSyncer                    syncers.SyncerInterface
+	downSyncer                  syncers.SyncerInterface
+}
+
+// Ready returns true if the controller is ready to return the GVRs to sync.
+// It implements [informer.GVRSource.Ready].
+func (c *controller) Ready() bool {
+	return c.syncConfigInformerHasSynced()
+}
+
+func (c *controller) enqueueSyncConfig(obj interface{}, logger klog.Logger) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	shared.WithQueueKey(logger, key).V(2).Info("queueing SyncConfig")
+	c.queue.Add(key)
+}
+
+// Start starts the controller workers.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := shared.WithReconciler(klog.FromContext(ctx), controllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
```

Comment:
This also is a "run" rather than a "start".

Reply:
Thanks. Fixed.